db/db_impl.cc

```diff
@@ -1029,6 +1029,15 @@ Status DBImpl::GetImpl(const ReadOptions& read_options,
   auto cfh = reinterpret_cast<ColumnFamilyHandleImpl*>(column_family);
   auto cfd = cfh->cfd();
 
+  if (tracer_) {
```

Comment:
According to Google search, unique_ptr is not thread safe. So perhaps it needs to be atomic pointer?

Reply:
I am using the Double-check locking idiom to avoid any overhead in the normal case when tracing is not enabled. https://en.wikipedia.org/wiki/Double-checked_locking .
In the case where tracing is enabled, we acquire a lock before tracing; so I believe it is thread safe. 

In the long run, we should change this logic to do multiple optimizations like batching the traces together, putting them on a queue, etc to avoid most of tracing overhead. But it is an optimization that I would not be able to get to until I see tracing being used more widely. Just trying to avoid pre-mature optimizations.

Reply:
What I mean is that the access of the pointer is not thread-safe --- it's OK for x86 but may not be the case for other platforms. TSAN will likely to complain anyway. I assume the straight-forward way of making this strictly allowed is to do

std::atomic<>. Maybe std::atomic<unique_ptr<>> will be compile to exactly the same as unique_ptr.

Reply:
Oh, I see; I misunderstood your first comment then.
Sure, let me test and make this atomic.

Reply:
A couple of runs of TSAN didn't fail on my machine. So, I will punt this issue for now, and circle back to it later as and when we work on more tracing optimizations.